tests/kernel/profiling/profiling_api/src/test_profiling_api.c

```diff
@@ -74,5 +82,96 @@ void test_call_stacks_analyze_workq(void)
 	}
 }
 
+static void thread_entry(void *p1, void *p2, void *p3)
+{
+	k_sleep(SLEEP_MS);
+}
+
+static void thread_analyze(struct k_thread *thread, void *user_data)
+{
+	int *var = (int *)user_data;
+
+	if (thread == &test_thread0) {
+		*var = 1;
+	}
+
+	thread_count++;
+	TC_PRINT("tid: %p, prio: %d\n", thread, thread->base.prio);
+}
+
+static void stack_analyze(const char *stack, size_t size, void *user_data)
+{
+	int *var = (int *)user_data;
+
+	if (stack == K_THREAD_STACK_BUFFER(test_stack1)) {
+		*var = 1;
+	}
+
+	thread_count++;
+	TC_PRINT("stack:%p, size:%u\n", stack, (unsigned int)size);
+}
+
+void test_k_thread_foreach(void)
+{
+	int temp;
+
+	/* Call k_thread_foreach() and check
+	 * thread_analyze is getting called.
+	 */
+	thread_count = 0;
+	temp = 0;
+	k_thread_foreach(thread_analyze, &temp);
+	zassert_true(thread_count && !temp,
+			"thread_analyze() not getting called");
+
+	TC_PRINT("Add new thread and call k_thread_foreach\n");
+	k_tid_t tid = k_thread_create(&test_thread0, test_stack0,
+			STACKSIZE, (k_thread_entry_t)thread_entry, NULL,
+			NULL, NULL, K_PRIO_PREEMPT(0), 0, 0);
+	k_sleep(1);
+
+	/* Call k_thread_foreach() and check
+	 * thread_analyze is getting called for
+	 * the newly added thread.
+	 */
+	thread_count = 0;
+	temp = 0;
+	k_thread_foreach(thread_analyze, &temp);
+	zassert_true(thread_count && temp,
+			"thread_analyze() not getting called");
+	k_thread_abort(tid);
+}
+
+void test_k_thread_stack_foreach(void)
+{
+	int temp;
+
+	/* Call k_thread_stack_foreach and check
+	 * stack_analyze is getting called.
+	 */
+	thread_count = 0;
+	temp = 0;
+	k_thread_stack_foreach(stack_analyze, &temp);
+	zassert_true(thread_count && !temp,
+			"stack_analyze() not getting called");
+
+	TC_PRINT("Add new thread and call k_thread_stack_foreach\n");
+	k_tid_t tid = k_thread_create(&test_thread1, test_stack1,
+			STACKSIZE, (k_thread_entry_t)thread_entry, NULL,
+			NULL, NULL, K_PRIO_PREEMPT(0), 0, 0);
+	k_sleep(1);
+
+	/* Call k_thread_stack_foreach and check
+	 * stack_analyze is getting called for
+	 * the newly added thread.
+	 */
+	thread_count = 0;
+	temp = 0;
+	k_thread_stack_foreach(stack_analyze, &temp);
+	zassert_true(thread_count && temp,
```

Comment:
This still doesn't prove that this API works.
It passes even if the callback only fires once and not for each thread in the system.

Reply:
@andrewboie the test is done in a relative way. I have called the API twice. In the first run, thread callback function counts the no.of threads present in the system. Then a new thread is being created. Now the k_thread_foreach() is again called. This time the thread count should be more than the previous one and we are also checking the tid is matching with newly added thread.

same approach is followed for k_thread_stack_foreach().

Note: I do not want to use the _kernel.threads to compare the tid's in the test's callback.